File: src/main/java/no/nav/pto/veilarbportefolje/oppfolging/SkjermingRepository.java
package no.nav.pto.veilarbportefolje.oppfolging;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.types.identer.Fnr;
import no.nav.pto.veilarbportefolje.oppfolging.response.SkjermingData;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.sql.ResultSet;
import java.sql.Timestamp;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import static no.nav.pto.veilarbportefolje.database.PostgresTable.NOM_SKJERMING.ER_SKJERMET;
import static no.nav.pto.veilarbportefolje.database.PostgresTable.NOM_SKJERMING.FNR;
import static no.nav.pto.veilarbportefolje.database.PostgresTable.NOM_SKJERMING.SKJERMET_FRA;
import static no.nav.pto.veilarbportefolje.database.PostgresTable.NOM_SKJERMING.SKJERMET_TIL;

@Slf4j
@Service
@RequiredArgsConstructor
public class SkjermingRepository {
    private final JdbcTemplate db;

    public boolean settSkjermingPeriode(Fnr fnr, Timestamp skjermetFra, Timestamp skjermetTil) {
        int updatedNum = db.update("""
                        INSERT INTO  NOM_SKJERMING(FODSELSNR, SKJERMET_FRA, SKJERMET_TIL) VALUES (?,?,?)
                        ON CONFLICT (FODSELSNR) DO UPDATE SET SKJERMET_FRA = EXCLUDED.SKJERMET_FRA, SKJERMET_TIL = EXCLUDED.SKJERMET_TIL
                        """,
                fnr.get(), skjermetFra, skjermetTil);
        return updatedNum > 0;
    }

    public boolean settSkjerming(Fnr fnr, Boolean erSkjermet) {
        int updatedNum = db.update("""
                        INSERT INTO NOM_SKJERMING (FODSELSNR, ER_SKJERMET) VALUES (?,?)
                        ON CONFLICT (FODSELSNR) DO UPDATE SET ER_SKJERMET = EXCLUDED.ER_SKJERMET
                        """,
                fnr.get(), erSkjermet);
        return updatedNum > 0;
    }

    public Optional<SkjermingData> hentSkjermingData(Fnr fnr) {
        return Optional.ofNullable(db.queryForObject("""
                        SELECT ER_SKJERMET, SKJERMET_FRA, SKJERMET_TIL FROM NOM_SKJERMING WHERE FODSELSNR = ?
                        """,
                (rs, rowNum) -> new SkjermingData(fnr, rs.getBoolean(ER_SKJERMET), rs.getTimestamp(SKJERMET_FRA), rs.getTimestamp(SKJERMET_TIL)), fnr.get())
        );
    }


    public Set<Fnr> hentSkjermetPersoner(List<Fnr> fnrs) {
        String fnrsCondition = fnrs.stream().map(Fnr::toString).collect(Collectors.joining(",", "{", "}"));
        Set<Fnr> skjermetPersoner = new HashSet<>();
        db.query("""
                        SELECT FODSELSNR FROM NOM_SKJERMING WHERE ER_SKJERMET AND FODSELSNR = ANY (?::varchar[])
                        """,
                ps -> ps.setString(1, fnrsCondition),
                (ResultSet rs) -> {
                    skjermetPersoner.add(Fnr.of(rs.getString(FNR)));
                }
        );
        return skjermetPersoner;
    }

    public void deleteSkjermingData(Fnr fnr) {
        db.update("""
                DELETE FROM NOM_SKJERMING WHERE FODSELSNR = ?
                """, fnr.get());
    }
}

File: src/main/java/no/nav/pto/veilarbportefolje/skjerming/SkjermingRepository.kt
package no.nav.pto.veilarbportefolje.skjerming

import no.nav.common.types.identer.Fnr
import no.nav.pto.veilarbportefolje.database.PostgresTable.NOM_SKJERMING
import org.jetbrains.annotations.TestOnly
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.queryForList
import org.springframework.stereotype.Service
import java.sql.ResultSet
import java.sql.Timestamp
import java.util.*

@Service
class SkjermingRepository(private val db: JdbcTemplate) {
    fun settSkjermingPeriode(fnr: Fnr, skjermetFra: Timestamp?, skjermetTil: Timestamp?): Boolean {
        val updatedNum = db.update(
            """
                INSERT INTO  NOM_SKJERMING(FODSELSNR, SKJERMET_FRA, SKJERMET_TIL) VALUES (?,?,?)
                ON CONFLICT (FODSELSNR) DO UPDATE SET SKJERMET_FRA = EXCLUDED.SKJERMET_FRA, SKJERMET_TIL = EXCLUDED.SKJERMET_TIL
                """.trimIndent(),
            fnr.get(), skjermetFra, skjermetTil
        )
        return updatedNum > 0
    }

    fun settSkjerming(fnr: Fnr, erSkjermet: Boolean): Boolean {
        val updatedNum = db.update(
            """
                INSERT INTO NOM_SKJERMING (FODSELSNR, ER_SKJERMET) VALUES (?,?)
                ON CONFLICT (FODSELSNR) DO UPDATE SET ER_SKJERMET = EXCLUDED.ER_SKJERMET
                """.trimIndent(),
            fnr.get(), erSkjermet
        )
        return updatedNum > 0
    }


    fun deleteSkjermingData(fnr: Fnr) {
        db.update(
            """
                DELETE FROM NOM_SKJERMING WHERE FODSELSNR = ?
                """.trimIndent(), fnr.get()
        )
    }

    @TestOnly
    fun hentSkjermingData(fnr: Fnr): Optional<SkjermingData> {
        return Optional.ofNullable(
            db.queryForObject(
                """
                    SELECT ER_SKJERMET, SKJERMET_FRA, SKJERMET_TIL FROM NOM_SKJERMING WHERE FODSELSNR = ?
                    """.trimIndent(),
                { rs: ResultSet, _: Int ->
                    SkjermingData(
                        fnr,
                        rs.getBoolean(NOM_SKJERMING.ER_SKJERMET),
                        rs.getTimestamp(NOM_SKJERMING.SKJERMET_FRA),
                        rs.getTimestamp(NOM_SKJERMING.SKJERMET_TIL)
                    )
                }, fnr.get()
            )
        )
    }

    @TestOnly
    fun hentSkjermetPersoner(fnrs: List<Fnr>): Set<Fnr> {
        val fnrsCondition = fnrs.joinToString(",", "{", "}", transform = Fnr::toString)

        return db.queryForList<String>(
            """
                SELECT FODSELSNR FROM NOM_SKJERMING WHERE ER_SKJERMET AND FODSELSNR = ANY (?::varchar[])
                """.trimIndent(),
            fnrsCondition
        ).map(Fnr::of).toSet()
    }
}
